sql/src/main/java/io/crate/action/sql/Session.java

```diff
@@ -376,49 +387,209 @@ public void execute(String portalName, int maxRows, ResultReceiver resultReceive
+                deferredExecutions.clear();
+                Portal portal = deferredExecution.portal();
                 LOGGER.debug("method=sync portal={}", portal);
-                pendingExecutions.clear();
-                clearState();
-                return portal.sync(planner, jobsLogs);
+                return singleExec(portal, deferredExecution.resultReceiver(), deferredExecution.maxRows());
+
             default:
-                throw new IllegalStateException(
-                    "Shouldn't have more than 1 pending execution. Got: " + pendingExecutions);
+                List<DeferredExecution> deferredExecutions = List.copyOf(this.deferredExecutions);
+                this.deferredExecutions.clear();
+                Map<Statement, List<DeferredExecution>> deferredExecutionsByStatement = deferredExecutions
+                    .stream()
+                    .collect(Collectors.toMap(
+                        (DeferredExecution x) -> x.portal().preparedStmt().parsedStatement(),
+                        List::of,
+                        Lists2::concat));
+                if (deferredExecutionsByStatement.size() == 1) {
+                    LOGGER.debug("method=sync bulkExec");
+                    return bulkExec(deferredExecutionsByStatement);
+                } else {
+                    LOGGER.debug("method=sync batchExec");
+                    List<? extends CompletableFuture<?>> futures = Lists2.map(
+                        deferredExecutions,
+                        x -> {
+                            if (!x.portal().boundOrUnboundStatement().isWriteOperation()) {
+                                throw new UnsupportedOperationException(
+                                    "Only write operations are allowed in Batch statements");
+                            }
+                            return singleExec(x.portal(), x.resultReceiver(), x.maxRows());
+                        }
+                    );
+                    return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
+                }
         }
     }
 
-    public void clearState() {
-        Portal portal = portals.remove(UNNAMED);
-        if (portal != null) {
-            portal.close();
+    private CompletableFuture<?> bulkExec(Map<Statement, List<DeferredExecution>> deferredExecutionsByStatement) {
+        var jobId = UUID.randomUUID();
+        var routingProvider = new RoutingProvider(Randomness.get().nextInt(), planner.getAwarenessAttributes());
+        var clusterState = executor.clusterService().state();
+        var txnCtx = new CoordinatorTxnCtx(sessionContext);
+        var plannerContext = new PlannerContext(
+            clusterState,
+            routingProvider,
+            jobId,
+            executor.functions(),
+            txnCtx,
+            0
+        );
+
+        var entry = deferredExecutionsByStatement.entrySet().iterator().next();
```

Comment:
shall we add an assertion on the size of the `deferredExecutionsByStatement` as we have a hard expectation of what it should be?

Reply:
I changed the signature to take the ` Statement` + `List<DeferredExecutions>` diretly. This way the unpacking of the entrySet is right after the `size` check.